File: cs325/midterm1/hanoi/hanoi.py
import os;
import sys;
import re;
import random;

def printTowers(A, B, C):
    print("src=%s" % (A))
    print("tmp=%s" % (B))
    print("dst=%s" % (C))

def initTowers(A, B, C):
    for i in range(0,int(sys.argv[1])):
        A.append(i+1);
        B.append(0);
        C.append(0);

def swap(src, dst, n):
    print("swap")
    dst.append(src.pop(0))
    dst.pop(0)
    src.insert(0,0); #syntax for inserting at index 0
    print("src=%s" % (src))
    print("dst=%s" % (dst))

def sortTowers(src, tmp, dst, n):
    if n == 0:
        return
    print("n=%s" % (n))
    sortTowers(src,dst,tmp,n-1)
    swap(src,dst,n)
    swap(tmp,dst,n)

def main():

    A = []
    B = []
    C = []

    initTowers(A, B, C)
    printTowers(A, B, C)
    sortTowers(A,B,C,len(A))
    printTowers(A, B, C)

main()

File: cs325/midterm1/hanoi/hanoi.cpp
#include <iostream>
#include <cstdlib>

using namespace std;

void printTower(int* T, int size)
{
	for (int i = 0; i < size; i++)
		cout << T[i] << ",";
	cout << endl;
}

void printA(int* A, int size)
{
	cout << "src: ";
	printTower(A,size);
}

void printB(int* B, int size)
{
	cout << "tmp: ";
	printTower(B,size);
}

void printC(int* C, int size)
{
	cout << "dst: ";
	printTower(C,size);
}

void printAll(int* A, int* B, int* C, int size)
{
	printA(A, size);
	printB(B, size);
	printC(C, size);
}

void initTowers(int* &A, int size)
{
	for (int i = 0; i < size; i++)
		A[i] = i+1;
}

void swap(int* &src, int* &dst, int size, int n)
//swaps idx-th elements of src and dst
{
	int low = 0;
	n--;
	//make low the index of the highest-index 0 in dst
	while (dst[low] == 0)
	{
		low++;
		if (low == size)
		{
			low--;
			break;
		}
		if (dst[low] != 0)
		{
			low--;
			break;
		}
	}	
	cout << "low=" << low << " n=" << n << endl;
	dst[low] = src[n];
	src[n] = 0;
	cout << endl;

}

void sortTowers(int* &src, int* &tmp, int* &dst, int size, int n)
{
	if (n == 0)
		return;
	sortTowers(src,dst,tmp,size,n-1);
	printAll(src,tmp,dst,size);
	swap(src,tmp,size,n);
	printAll(src,tmp,dst,size);
	swap(tmp,dst,size,n);
}

int main(int argc, char* argv[])
{
	if (argc == 1)
		return 0;
	int size = atoi(argv[1]);
	cout << size << endl;
	
	int* A = new int[size];
	int* B = new int[size];
	int* C = new int[size];

	initTowers(A, size);

	sortTowers(A, B, C, size, size);
	
	printAll(A, B, C, size);
	
	delete(C);
	delete(B);
	delete(A);
}
